clippy_lints/src/derive.rs

```diff
@@ -419,3 +453,43 @@ impl<'tcx> Visitor<'tcx> for UnsafeVisitor<'_, 'tcx> {
         self.cx.tcx.hir()
     }
 }
+
+/// Implementation of the `DERIVE_PARTIAL_EQ_WITHOUT_EQ` lint.
+fn check_partial_eq_without_eq<'tcx>(
+    cx: &LateContext<'tcx>,
+    span: Span,
+    trait_ref: &TraitRef<'_>,
+    ty: Ty<'tcx>,
+    peq_is_automatically_derived: bool,
+) {
+    if_chain! {
+        if let ty::Adt(adt, substs) = ty.kind();
+        if peq_is_automatically_derived;
+        if let Some(eq_trait_def_id) = cx.tcx.get_diagnostic_item(sym::Eq);
+        if let Some(def_id) = trait_ref.trait_def_id();
+        if cx.tcx.is_diagnostic_item(sym::PartialEq, def_id);
+        then {
+            if !implements_trait(cx, ty, eq_trait_def_id, substs) {
+                // If all of our fields implement `Eq`, we can implement `Eq` too
+                for variant in adt.variants() {
```

Comment:
This section is based off [can_type_implement_copy](https://github.com/rust-lang/rust/blob/37b55c8a0cafdb60b9168da34f904acc70157df8/compiler/rustc_trait_selection/src/traits/misc.rs#L19) from rustc. Only thing I wasn't super sure about is that that function does a [check](https://github.com/rust-lang/rust/blob/37b55c8a0cafdb60b9168da34f904acc70157df8/compiler/rustc_trait_selection/src/traits/misc.rs#L490) for `ty.references_error()`, it wasn't sure clear to me if that was necessary here (or indeed what exactly it does)

Reply:
I'm fairly sure you can ignore that. By the time late lint passes run the code already has no errors.